### PR TITLE
fix: normalize nested field nullability in ShuffleScanExec and ExpandExec

### LINKED ISSUE #5137: ShuffleScanExec and ExpandExec assert on nested field nullability instead of normalizing it
### Describe the bug

Comet has two independent sources of truth for a column's Arrow type:

1. **Declared** — Spark catalyst's type, serialized per nested field over protobuf (`StructInfo.field_nullable`, `ListInfo.contains_null` in `native/proto/src/proto/types.proto`) and rebuilt by `to_arrow_datatype` (`native/core/src/execution/serde.rs`).
2. **Actual** — the Arrow type of the array a kernel, scan, or shuffle block really produced.

Arrow treats nested field nullability as part of `DataType` identity: `DataType::equals_datatype` compares `is_nullable()` at every list element, struct field, and map entry, and `RecordBatch::try_new_with_options` has no option to relax it (`match_field_names: false` only ignores field *names*). So any operator that **stamps** the declared schema onto a runtime array hard-fails when the two disagree by a single nested `nullable` flag, e.g.:

```
Invalid argument error: column types must match schema types,
expected List(Struct("id": Int64, "flag": Boolean))
but found List(Struct("id": Int64, "flag": non-null Boolean))
at column index 0
```

The data is fine in this situation — a non-nullable child is a strict subset of a nullable one, and the batch would produce identical results. It is a metadata assertion failure, not corruption.

Most boundaries already **normalize** instead of asserting, and that is why this drift is usually invisible:

- `ScanExec::build_record_batch` (`native/core/src/execution/operators/scan.rs`) casts every imported column to the declared type. Arrow's cast rewrites nested fields from the target (`cast_struct_to_struct` builds with `to_fields`, `cast_list_values` with `to.clone()`), so the FFI boundary absorbs nullability drift.
- `SchemaAlignExec` (`native/shuffle/src/schema_align.rs`) casts the shuffle writer's input, added for #4515.
- `CometLocalTableScanExec` widens its output with `.asNullable`, added for #4789.

Two stamp points still assert:

- `ShuffleScanExec::poll_next` (`native/core/src/execution/operators/shuffle_scan.rs`) — stamps the catalyst schema onto the decoded shuffle block with no cast. This is the direct-read shuffle path, enabled by default (`spark.comet.shuffle.directRead.enabled`).
- `ExpandStream::expand` (`native/core/src/execution/operators/expand.rs`) — the schema is derived from `projections[0]` only (see `OpStruct::Expand` in `native/core/src/execution/planner.rs`) and then stamped onto the output of *every* projection, so the operator fails if any two projections disagree on a nested nullable flag for the same output column.

Because the surviving stamp points are both inside native plans (past `ScanExec`'s cast), the failure mode is: a native kernel or non-Parquet source produces a nested type that is narrower than what catalyst declared, nothing corrects it, and a downstream `ShuffleScanExec` or `ExpandExec` aborts the task. Retries do not help; it is deterministic for the partition's data shape.

Upstream nested-nullability drift is tracked by #4515 (with concrete kernel instances in #4789 and #4528). This issue is about the boundaries: the same drift should be absorbed rather than fatal, the way the other boundaries already do.

A secondary problem: the arrow error text names neither the operator nor the column name, only `at column index N`. For a wide schema of deeply nested structs this makes the report very hard to act on, since the two printed types can differ by one flag hundreds of characters in.

### Steps to reproduce

Not yet reduced to a standalone test. Both sites need a native plan in which a column's actual nested nullability differs from the catalyst-declared type:

- `ShuffleScanExec`: a native plan reading shuffle output where the block's nested field nullability differs from the consuming stage's declared type.
- `ExpandExec`: a grouping-sets / rollup / multiple-distinct-aggregate plan over an `array<struct<...>>` column where one projection's expression yields a different nested nullable flag than `projections[0]`.

Writing the tests is part of the work here. The `SchemaAlignExec` unit tests are a good model — they pin each drift pair so the test flips to a passthrough assertion once the upstream type is corrected.

### Expected behavior

Neither boundary should fail on a nested nullability difference:

- `ShuffleScanExec` should reconcile the decoded block with the declared schema the way `ScanExec` does (cast per column, or reuse `SchemaAlignExec`'s alignment logic on the read side).
- `ExpandExec` should build its output schema so that it is valid for every projection, not just `projections[0]` — e.g. widen nested nullability across all projections, or align each projection's output to the operator's declared schema before stamping.

Additionally, the error raised when a stamp genuinely cannot be reconciled should name the operator and the offending column path, not just the column index.

### Additional context

Related: #4515 (epic: return-type drift from DataFusion / datafusion-spark), #4789 (non-null nested child fields from local table scans), #4528 (`make_array` with nullability-divergent struct children).

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes #5137.

## Rationale for this change

Arrow treats nested field nullability as part of `DataType` identity, so `RecordBatch::try_new_with_options` rejects a column whose nested `nullable` flags are narrower than the declared type — even though the data is fine, since a non-null child is a strict subset of a nullable one. It is a metadata assertion failure, not corruption.

Most Comet boundaries already normalize instead of asserting, which is why this drift is usually invisible:

- `ScanExec::build_record_batch` casts every imported column to the declared type, so the FFI boundary absorbs it.
- `SchemaAlignExec` casts the shuffle writer's input (added for #4515).
- `CometLocalTableScanExec` widens its output with `.asNullable` (added for #4789).

Two stamp points still asserted, and because both sit inside native plans (past `ScanExec`'s cast) nothing corrected the drift before them. The task aborted deterministically for the partition's data shape, so retries did not help:

```
Invalid argument error: column types must match schema types,
expected List(Struct("id": Int64, "flag": Boolean))
but found List(Struct("id": Int64, "flag": non-null Boolean))
at column index 0
```

- `ShuffleScanExec::poll_next` stamped the catalyst schema onto the decoded shuffle block with no cast. This is the direct-read shuffle path, enabled by default (`spark.comet.shuffle.directRead.enabled`).
- `ExpandStream::expand` derived its schema from `projections[0]` only and stamped it onto the output of *every* projection, so it failed if any two projections disagreed on a nested nullable flag for the same output column.

## What changes are included in this PR?

A new `cast_and_stamp_schema` helper in `datafusion-comet-common`, used at both stamp points, which casts any column whose Arrow type differs from the declared field type before stamping. Arrow's cast rewrites nested fields from the target (`cast_struct_to_struct` builds with `to_fields`, `cast_list_values` with `to.clone()`), which is exactly how `ScanExec` already absorbs this.

`ExpandExec::build_schema` now derives the output schema by widening nested nullability across **all** projections instead of reading it off `projections[0]`, so the declared schema is valid for every projection. The runtime cast is a second, independent line of defence: each is tested on its own, so neither silently carries the other.

Error messages now name the operator and the dotted path of the offending column:

```
CometExpandExec cannot reconcile col[0] with its declared schema at
col_0.element.flag: expected nullable Boolean, found non-null Boolean
```

instead of only `at column index N`, which for a wide schema of deeply nested structs was very hard to act on — the two printed types could differ by one flag hundreds of characters in.

A narrowing cast (declared non-null child, actual data contains nulls) is not silently accepted: `StructArray::try_new` rejects unmasked nulls for a non-nullable field, so that case still errors loudly, now with the column path named.

This PR is about the boundaries. The upstream nested-nullability drift itself remains tracked by #4515, with concrete kernel instances in #4789 and #4528.

## How are these changes tested?

New Rust unit tests, modelled on the existing `SchemaAlignExec` tests as the issue suggested:

- `native/common/src/schema.rs` — 9 tests over `cast_and_stamp_schema`, `describe_type_mismatch`, and `widen_nested_nullability`. `stamps_nested_nullability_drift` first asserts that the plain stamp *does* reject the drift, then that the helper absorbs it with values intact. That assertion pins why the helper exists: if arrow ever relaxes nested nullability comparison, it flips and the call sites can go back to a plain stamp.
- `native/core/src/execution/operators/expand.rs` — 5 tests: `build_schema` widens across projections (driven from the narrow side, so it cannot pass by echoing `projections[0]`), rejects projections of differing width, expands divergent projections end-to-end, and reconciles even when handed the narrow `projections[0]`-derived schema the planner used to build.
- `native/core/src/execution/operators/shuffle_scan.rs` — 2 tests: a decoded block with a non-null nested child against a nullable declared type is reconciled with values preserved, and an unreconcilable column names the operator and column.

I confirmed these are genuine regression tests rather than tests that pass either way: without the fix, `RecordBatch::try_new_with_options` rejects the drift, which the first assertion in `stamps_nested_nullability_drift` now pins permanently.

Also run and passing: full `cargo test --workspace` (135 + 33 native tests; the 3 pre-existing `datafusion-comet-jni-bridge` `errors::tests` panic-handling failures reproduce on a clean checkout of this branch point and are unrelated), `cargo clippy --all-targets --workspace` clean, `cargo fmt`, and on Spark 3.5: `CometExecSuite "expand operator"`, `CometNativeShuffleSuite` (27), `CometShuffleSuite` (44), `CometAggregateSuite` (85, 2 canceled).

I did not reduce this to a Spark-level SQL test. Both sites need a native plan where a kernel actually produces a nested type narrower than catalyst declared, and those kernels are the subject of #4515/#4789/#4528 — so a SQL repro would be pinned to whichever kernel currently drifts and would evaporate when that kernel is fixed, which is the wrong thing to encode. The tests above pin the boundaries directly, independent of which kernel drifts.

## Are these changes tested?

Yes.

## Are there any user-facing changes?

No API changes. Queries that previously aborted with `column types must match schema types` in the direct-read shuffle path or in Expand (grouping sets / rollup / multiple distinct aggregates over nested columns) now complete. Errors that remain genuinely unreconcilable are reported with the operator and column path named.

## Follow-up

`ScanExec::build_record_batch` does the same cast-then-stamp but keeps its own copy of the logic and its own `cast_time` metric, so its errors still report only `at column index N`. Switching it to the shared helper would need the helper to accept a timer; I left it out to keep this diff focused on the two boundaries the issue names.

### CODE REVIEW

File: native/core/src/execution/operators/expand.rs [hunk 1]
```diff
@@ -65,6 +66,47 @@ impl ExpandExec {
...
+            }
+            for (data_type, expr) in data_types.iter_mut().zip(projection) {
+                *data_type = widen_nested_nullability(data_type, &expr.data_type(input_schema)?);
+            }
+        }
+
+        let fields: Vec<Field> = data_types
+            .into_iter()
+            .enumerate()
+            .map(|(idx, dt)| Field::new(format!("col_{idx}"), dt, true))
```
Comment (hsiang-c) [inline native/core/src/execution/operators/expand.rs:110]: Is it required to keep the field name as is?

Comment (andygrove) [reply native/core/src/execution/operators/expand.rs:110]: Not strictly required for correctness, but I kept it deliberately.

`col_{idx}` is exactly what the planner produced before this logic moved into `build_schema`, so keeping it means the only behavior change here is the nullability derivation. There is also no better name available: an Expand projection is an arbitrary `PhysicalExpr` (including literals like the grouping ID), so there is nothing to inherit — and parent operators bind to these columns by index, with a bound reference taking its name from the child schema (`planner.rs`, `ExprStruct::Bound`), so the name is positional either way. It is only visible in explain output.

Added a comment at the call site recording that.

**Review by hsiang-c — APPROVED** [review]
Thanks @andygrove LGTM

File: native/common/src/schema.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+/// messages.
+///
+/// This is the normalizing counterpart to stamping `schema` on directly: it absorbs the
+/// return-type drift — most commonly a nested `nullable` flag — that native kernels and non-Parquet
+/// sources introduce, the same way `ScanExec` absorbs it at the FFI boundary.
+///
+/// Reconciliation follows `schema` in both directions, so it will also narrow a nullable nested
+/// child to non-null when that is what `schema` declares. That is not silently lossy: arrow's
+/// `StructArray::try_new` rejects unmasked nulls under a non-nullable field, so data that cannot
+/// survive the narrowing errors here rather than producing an array that misreports itself.
```
Comment (parthchandra) [inline native/common/src/schema.rs:46]: I don't think this is covered in the test. Maybe add a test that stamps a nullable-child array carrying a real null against a non-null-declared schema and asserts it errors?

**Review by parthchandra — APPROVED** [review]
lgtm. One small suggestion (not blocking though).

File: native/common/src/schema.rs (hunk 2 above)
Comment (andygrove) [reply native/common/src/schema.rs:46]: Thanks, will do as follow up. Merging this to unblock #5051

Comment (andygrove) [reply native/common/src/schema.rs:46]: Good catch, that claim was untested. Added in https://github.com/apache/datafusion-comet/pull/5285 — it stamps a nullable-child array against a non-null-declared schema both with and without a real null, so the child's `nullable` flag alone does not decide the outcome: the null-free case narrows fine, and the one carrying a null errors with the `c0.element.flag` path named.
